### DISCUSSION TITLE
When two document types belongs to different schemas having same alias not allowed. Giving ambiguity error when accessed using IDocumentSession.

### DISCUSSION THREAD

**Discussion by maheshnicham123** [JasperFx/marten#2948] [Q&A]
Hi Team,

I have scenario where 2 document types are having same alias(or classnames), but they belongs to different schemas like below -
[Same database]

**First document type -:** 
[DatabaseSchemaName("public")]
public class MaterialEntity : BaseEntity
{
//Properties specific to this document goes here.
}

**Second document type -:** 
[DatabaseSchemaName("manufacturing")]
public class MaterialEntity : AggregateBaseEntity
{
//Properties specific to this document goes here.
}

**Trying to fetch first document type using it's repository below -:**

public class MaterialRepository : IMaterialRepository
{
    private readonly IDocumentSession _dbSession;

    public MaterialRepository(IDocumentSession dbSession)
    {
        _dbSession = dbSession ?? throw new ArgumentNullException(nameof(dbSession));        
    }

    public async Task<MaterialEntity> GetMaterialById(string id)
    {
        return await _dbSession.Query<MaterialEntity>().FirstOrDefaultAsync(x => x.SapId == id);       
    }
}

Similarly accessing second document type using it's specific repository.

**Below marten library code is giving AmbiguousDocumentTypeAliasesException** 
[Using Marten 5.11.0 nuget]

private void assertNoDuplicateDocumentAliases()
        {
            var duplicates =
                AllDocumentMappings.Where(x => !x.StructuralTyped)
                    .GroupBy(x => x.Alias)
                    .Where(x => x.Count() > 1)
                    .ToArray();
            if (duplicates.Any())
            {
                var message = duplicates.Select(group =>
                {
                    return
                        $"Document types {group.Select(x => x.DocumentType.FullName!).Join(", ")} all have the same document alias '{group.Key}'. You must explicitly make document type aliases to disambiguate the database schema objects";
                }).Join("\n");

                throw new AmbiguousDocumentTypeAliasesException(message);
            }
        }

I know this issue can be fixed by changing the document type class name. But business is not allowing it because there will be more document types added with similar name in future.

Further I tried to group document types by there schema using below DI( It worked) but now I am unable to implement UoW using TransactionScope object (I am using Meidar library and there are nested calls to handler-pipeline-behavior)

//This dependency injection for creating documentsession/querysession specific to 'public' schema document types
        builder.Services.AddMartenStore<ISynchronizationStore>(opts =>
        {
            opts.Connection(dbConnectionString);
            opts.DatabaseSchemaName = "public";

        });

        //This dependency injection for creating documentsession/querysession specific to 'manufacturing' schema document types
        builder.Services.AddMartenStore<IManufacturingStore>(opts =>
        {
            opts.Connection(dbConnectionString);
            opts.DatabaseSchemaName = "manufacturing";
        });

Is there any way to instruct marten to use fully qualified name of documenttype instead of alias only?
Please advise!

**Comment by jeremydmiller** [comment]
See this: https://martendb.io/documents/storage.html#type-aliases

**Reply by maheshnicham123** [reply]
Hi @jeremydmiller ,

Thanks for your response!

I know this solution(using [DocumentAlias] attribute or MartenRegistry) and already tried but we don't want change document type name in db.

e.g.

[DatabaseSchemaName("public")]
[DocumentAlias("publicmaterial")]
public class MaterialEntity : BaseEntity
{
//Properties specific to this document goes here.
}

This will auto create db entity like '.mt_doc_publicmaterial' under 'public' schema but I want '.mt_doc_materialentity' which is already in production.

[DatabaseSchemaName("manufacturing")]
[DocumentAlias("manufacturingmaterial")]
public class MaterialEntity : AggregateBaseEntity
{
//Properties specific to this document goes here.
}

This will auto create db entity like '.mt_doc_manufacturingmaterial' under 'manufacturing' schema but I want '.mt_doc_materialentity' which is already in production.


And these document types should be accessed like below query from .Net Core.
[Query works in PgAdmin]

SELECT id, data, mt_last_modified, mt_version, mt_dotnet_type
	FROM public.mt_doc_materialentity;

Any advise on this please?

**Reply by maheshnicham123** [reply]
Hi @jeremydmiller,

Any solution on this scenario please?

Regards,
Mahesh Nicham
